iothub_client/tests/common_dt_e2e/iothubclient_common_dt_e2e.c

```diff
@@ -586,390 +746,238 @@ void client_create_with_properies_and_send_d2c(IOTHUB_PROVISIONED_DEVICE* device
-                        {
-                            Unlock(device->lock);
-                            break;
-                        }
-                    }
-                }
-                json_value_free(root_value);
-                root_value = NULL;
-            }
-            Unlock(device->lock);
-        }
-        ThreadAPI_Sleep(1000);
-    }
+    _create_client_handle(device_to_use, protocol);
 
-    ASSERT_IS_TRUE(difftime(nowTime, beginOperation) < MAX_CLOUD_TRAVEL_TIME, "Timeout waiting for twin message");
+    // Twin registrations to the cloud happen asyncronously because we're using the convenience
+    // layer. There is an (unavoidable) race potential in tests like this where we create handles
+    // and immediately invoke the service SDK. Namely we could:
+    // 1 - Register for the full twin (which happens via IoTHubDeviceClient_SetDeviceTwinCallback)
+    // 2 - Have the service SDK update the twin (see _service_client_update_twin), but it takes a while
+    // 3 - The client receives its full twin, which will just be empty data given (2) isn't completed
+    // 4 - When the client receives full twin, it will register for PATCH changes
+    // 5 - The server only now completes (2), setting the full twin.  However this has happened
+    //     *after* it received the subscribe for PATCH and therefore it doesn't send down the PATCH
+    //     of the full twin.
+    // Apps in field will rarely hit this, as it requries service SDK & client handle to be invoked
+    // almost simultaneously.  And the client *is* registered for future twin updates on this handle,
+    // so it would get future changes.
+    //
+    // To ensure the test passes: Receive first full twin in the test app. Then sleep so the server
+    // can receive the subscribe for PATCH. Then have the service sdk update the twin on the Hub.
 
-    // unsubscribe
-    setdevicetwincallback_on_device_or_module(NULL, NULL);
+    // Connect device client to IoT Hub. Register callback. Receive full twin.
+    _set_twin_callback(_twin_callback, received_twin_data);
+    _receive_twin_loop(received_twin_data, DEVICE_TWIN_UPDATE_COMPLETE);
+    _received_twin_data_reset(received_twin_data);
 
-    if (Lock(device->lock) != LOCK_OK)
-    {
-        ASSERT_FAIL("Lock failed");
-    }
-    else
-    {
-        ASSERT_IS_NOT_NULL(root_value, "json_parse_string failed");
-        ASSERT_ARE_EQUAL(char_ptr, expected_desired_string, string_property, "string data retrieved differs from expected");
-        ASSERT_ARE_EQUAL(int, expected_desired_integer, integer_property, "integer data retrieved differs from expected");
-        ASSERT_ARE_EQUAL(char_ptr, expected_desired_string, string_property_from_array, "string data (from array) retrieved differs from expected");
-        ASSERT_ARE_EQUAL(int, expected_desired_integer, integer_property_from_array, "integer data (from array) retrieved differs from expected");
+    ThreadAPI_Sleep(5000);
```

Comment:
5000 is magic # in code - could we put it as a `static const unsigned int` at top of file and the places that reference it reference that one instead?